sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala

```diff
@@ -92,6 +93,67 @@ class JoinSuite extends QueryTest with SharedSparkSession with AdaptiveSparkPlan
     operators.head
   }
 
+  test("NO_BROADCAST_AND_REPLICATION hint is respected in cross joins") {
```

Comment:
These 3 tests cover scenarios when it is not safe to broadcast or replicate the target table to perform the cardinality check. The newly added internal hint handles this. There are MERGE tests for this too.